### ISSUE TITLE
Need for implementing .unique() for arrays

### ISSUE THREAD

**Issue by amirhossein-razlighi** [ml-explore/mlx#856]
Hi. I think a `.unique()` function is needed for `mx.array` types, to return number of unique items in a `mx.array` tensor. This also makes the framework more compatible to _numpy_ and _pytorch_ conventions related to this task.
If this is a valid issue, please let me know. I am willing to work on this myself, if possible.
Thanks!

**Comment by awni** [comment]
`unique` returns the set of unique elements, not the number. This is actually a difficult op to implement in MLX because it falls in the category of ops whose outputs shape depends on input data (like boolean masking, unary where, etc).

For now I would say we will not implement it until we have figure out a consistent strategy for handling such ops.

**Comment by yaroslavyaroslav** [comment]
No updates here? 

Although what the one could suggest to use for such operations until you'd figure out such strategy? LIke casting back and forth to torch and mlx?

**Comment by awni** [comment]
Cast to numpy is probably better. But I would encourage you to see if you can work around the need for unique entirely. If you can share a simplified version of your use case (some code) maybe we can make a suggestion.

**Comment by yaroslavyaroslav** [comment]
I believe this is quite ultimate example that also belongs on `where` in addition to `unique`.

Would really appreciate to get help with that.

```py
def row_difference(A, B: torch.tensor) -> torch.tensor:
    r"""
        Calulates A \ B — set difference between rows of A and rows of B
    """
    _, idx, counts = torch.cat([A, B], dim=0).unique(dim=0, return_inverse=True, return_counts=True)
    mask = torch.isin(idx, torch.where(counts.gt(1))[0])
    return A[~mask[:len(A)]]
```

**Comment by amirhossein-razlighi** [comment]
A simple workaround (maybe it can be optimized further, it's just a sample!):

```
import torch
import mlx.core as mx
import numpy as np

def function_torch(A, B):
    _, idx, counts = torch.cat([A, B], dim=0).unique(
        dim=0, return_inverse=True, return_counts=True
    )
    mask = torch.isin(idx, torch.where(counts.gt(1))[0])
    return A[~mask[: len(A)]]


def function_mlx(A, B):
    A = mx.array(A)
    B = mx.array(B)
    cated = mx.concatenate([A, B], axis=0)
    element_to_idx = {}
    counts = mx.zeros(len(cated), dtype=mx.int32)

    counter = 0
    for i in range(len(cated)):
        e = cated[i].item()
        if e in element_to_idx:
            counts[element_to_idx[e]] += 1
        else:
            element_to_idx[e] = counter
            counts[counter] = 1
            counter += 1
    
    counts = counts[:counter]
    mask = counts > 1
    return mx.array(np.array(A)[mx.logical_not(mask[:len(A)])])


A = torch.tensor([1, 2, 3, 4, 5])
B = torch.tensor([1, 2, 3, 6, 7])
print(function_torch(A, B))
print(function_mlx(A, B))
```

**Comment by amirhossein-razlighi** [comment]
The problem is mlx doesn't have boolean indexing for now. So, I had to covet to numpy array in the end to be able to do boolean indexing and convert back to mlx array.

**Comment by awni** [comment]
If you can use a mask instead of the actual row difference that is much better and should be doable without needing `unique` or unary `where`. I would recommend you take a look [at this guide](https://gist.github.com/awni/e20a288011adac0d0bdaf73e5296ca4f) for more insight there.

If it's not possible to use the mask .. then I would convert to Numpy (just once) do the op and convert back if needed.

**Comment by yaroslavyaroslav** [comment]
@amirhossein-razlighi thanks for suggestion, it helped a lot!
